### PR TITLE
New developer doc for adding new feature

### PR DESCRIPTION
**Pull request by scsides**
Add new document for developers about adding new features

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [X] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [X] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [ ] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [X] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

### CODE REVIEW

File: isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+            program in ISIS is focussed on a narrow subject, for example, isis2std is focussed on exporting ISIS cubes
+            to standard image formats such as jpeg, tif, or png.
+          </p>
+          <p>
+            ISIS applications are written in C++ and are supported on Ubuntu Linux and Mac OS.
+          </p>
+
+        <h2><a name="Resources">Resources</a></h2>
+          <p>
+            The <a href="https://isis.astrogeology.usgs.gov/Object/Developer/index.html>ISIS library"</a> contains
```
Comment (jessemapel) [inline isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml]: Make this a relative path.

File: isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+            starting to write your first application.
+          </p>
+          <p>
+            There are over 400 <a href="https://isis.astrogeology.usgs.gov/Application/index.html"ISIS applications</a>.
+            We suggest you get familiar with as many of these as possible, so you don't duplicate a feature that has already 
+            been implemented, and to use an existing application as a template for your new feature.
+          </p>
+          <p>
+            ISIS applications are linked against many external libraries. We encourage you to become familiar with these 
+            external libraries. One place to start is the environment.yml file located at the root of the ISIS source
```
Comment (jessemapel) [inline isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml]: I would say that "a list of the external libraries that ISIS uses directly can be found in the environment.yml file"

File: isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+              <a href="https://github.com/USGS-Astrogeology/ISIS3/wiki/Developing-ISIS3-with-cmake#building-isis3">compile and build</a> 
+              all of ISIS. This step will make sure sure you have everything setup and ready to go.
+            </p>
+
+          <h2><a name="Starting a conversation with other ISIS developers">Starting a conversation with other ISIS developers</a></h2>
+            <p>
+              Before you start writing your new feature, create an <a href="https://github.com/USGS-Astrogeology/ISIS3">issue</a> for 
+              it on the ISIS GitHub site. Doing this will allow other ISIS developers to way in on your preposed feature. By posting
+              a description of what you want to do and asking questions, other developers can give you pointers on how to 
+              implement it.
```
Comment (jessemapel) [inline isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml]: Add a note that the new feature should be labeled as an enhancement.

File: isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+            <p>
+              The ISIS source code is organized by having a dedicated directory where all of the source code for an application is
+              stored. The files in this directory will be exclusively to your feature. This is where using another application as
+              a template can be very useful.
+            </p>
+            <ol>
+                <li>Decide where your new application belongs</li>
+                <li>Create a directory using the same name as you new application</li>
+                <li>Create a "main.cpp" file</li>
+                <li>Create a "your_new_app.cpp" file</li>
```
Comment (jessemapel) [inline isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml]: You will also need a header file for the app

File: isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+                <li>Create a "your_new_app.xml" file</li>
+                <li>Create any additional support files (e.g., support_class.cpp, support_class.h</li>
```
Comment (jessemapel) [inline isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml]: Change these filenames to be all lowercase for app and upper camel case for classes to match code style.

Comment (jessemapel) [inline isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml]: I would encourage developers to also create test files at this point. One for the app and one for each supporting class.

File: isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+            </ol>
+            <p>
+              Populate the new files with the code necessary to implement your feature.
```
Comment (jessemapel) [inline isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml:107]: We should give instructions or link to the instructions for writing app mains and app functions.

File: isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+            </p>
+
+            <h3>Compile and link your new application for the first time</h3>
+            <p>
+              Since you are adding a new application to ISIS you will need to re-run the cmake command. This will add
```
Comment (jessemapel) [inline isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml]: "re-run the cmake command" -> "re-configure your build with cmake"

File: isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+            </p>
+
+            <h3>Compile and link your new application for the first time</h3>
+            <p>
+              Since you are adding a new application to ISIS you will need to re-run the cmake command. This will add 
+              the new application to the list of applications the build system knows about.
+            </p>
+            <ol>
+              <li>Change you current working directory to the ISIS build directory (cd $ISISROOT)</li>
+              <li>Remove the contents of the build directory (rm -rf *)</li>
```
Comment (jessemapel) [inline isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml]: We don't need to go this far. You can just remove `CMakeCache.txt` and it should re-do everything but retain the already built files. It'll save a lot of time.

File: isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+              <li>Rebuild the ISIS library and your new application (ninja)</li>
+              <li>Run your application and test the user interface</li>
+              <li>Run your application and test the output</li>
+                <ol>
+                  <li>Cube DNs</li>
+                  <li>Cube labels</li>
+                  <li>History</li>
+                  <li>Original label</li>
+                  <li>print.prt file</li>
+                  <li>Cleanup any temporary files</li>
```
Comment (jessemapel) [inline isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml]: What does this list mean? What should developers be checking for here? Being more specific like "Does the app add a history entry to all output cubes?"

File: isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+                </ol>
+            </ol>
+
+            <h2><a name="Writing your new application">Writing your new application</a></h2>
```
Comment (jessemapel) [inline isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml]: This looks like it should be Testing your new application

File: isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+              Applications in ISIS are required to have tests that exercise the majority of the application specific code.
+              These test are written using Google Test (aka, gtest). The source code for all gtests is located in the "isis/tests" directory
```
Comment (jessemapel) [inline isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml]: Google Test -> googletest

File: isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+              starting at the root of the code repository. Create a new file for your new application with the name "FunctionalTestsYournewapp.cpp", and
+              write tests to exercise the application. Remember to test only the functionality of your new code and not the ISIS library. Tests for the
+              ISIS library already exist.
```
Comment (jessemapel) [inline isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml]: Mention that you should unit test any support classes that are sufficiently complex.

Comment (jessemapel) [reply isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml]: Also link to the testing docs here

File: isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+          <ol>
+            <li>Tell git to track your new files (git add new_files)</li>
+            <li>Commit them to your local repository (git commit -m "brief description")</li>
+            <li>Push the new application to your GitHub fork (git push origin)</li>
+            <li>Create a pull request on <a href="https://github.com/USGS-Astrogeology/ISIS3/compare">GitHub</a></li>
+            <li>Iterate with the reviewers to fix any issues, resolve and confilcts and explain any questions</li>
+            <li>Update your pull request with any changes</li>
+          </ol>
+          <p>
+            Once your reviewers are satisfied one of them will merge your new application code with the ISIS dev brand. When
```
Comment (jessemapel) [inline isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml]: brand -> branch

File: isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+              ISIS library already exist.
+            </p>
+
+          <h2><a name="Submit your new application for review">Submit your new application for review</a></h2>
+          <p>
+            Now that you have all the pieces for your new application and it works perfectly. It's time to get it reviewed by
+            other ISIS developers, so it can be added to ISIS and be distributed with the next public feature release.
+          </p>
+          <ol>
+            <li>Tell git to track your new files (git add new_files)</li>
```
Comment (jessemapel) [inline isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml:175]: Add a changelog entry
Add yourself to the zenodo file

**Review by jessemapel — COMMENTED** [review]
Update the link in DevStart.xsl to point to this.

File: isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+              <li> Editor (Vim, Nano, Atom, ...</li>
+            </ul>
+
+            <p>
+              Once you have a complete environment set up for ISIS development, we suggest you 
+              <a href="https://github.com/USGS-Astrogeology/ISIS3/wiki/Developing-ISIS3-with-cmake#building-isis3">compile and build</a> 
+              all of ISIS. This step will make sure sure you have everything setup and ready to go.
+            </p>
+
+          <h2><a name="Starting a conversation with other ISIS developers">Starting a conversation with other ISIS developers</a></h2>
```
Comment (jessemapel) [inline isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml:74]: add a link to the contributing document that outlines expectations of contributions and interactions https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md

Comment (scsides) [reply isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml:74]: Placed nearer the top

File: isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml (hunk 8 above)
Comment (scsides) [reply isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml]: Good point

File: isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml [hunk 16]
```diff
@@ -118,22 +132,39 @@
...
-                  <li>History</li>
-                  <li>Original label</li>
-                  <li>print.prt file</li>
-                  <li>Cleanup any temporary files</li>
+                  <li>Cube DN's - Are they what you expect?</li>
+                  <li>Cube labels - Have they been modified as you intend?</li>
+                  <li>History - Did the application add a cube history entry for every output cube?</li>
+                  <li>Original label - If your application is an import program, did it store the original label in the output cube</li>
+                  <li>print.prt file - Did your application write a log entry to the print.prt file?</li>
+                  <li>Cleanup temporary files - If you application used any temporary files, did it remove them before it exited?</li>
```
Comment (jessemapel) [inline isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml]: If you**r** application

File: isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+            </p>
+
+            <ol>
+                <li>Decide where your new application belongs</li>
+                <li>Create a directory using the same name as you new application</li>
+                <li>Create a "main.cpp" file</li>
+                <li>Create a "your_new_app.cpp" file</li>
+                <li>Create a "your_new_app.h" file</li>
+                <li>Create a "your_new_app.xml" file</li>
+                <li>Create any additional support files (e.g., Support_class.cpp, Support_class.h</li>
```
Comment (jessemapel) [inline isis/src/docsys/documents/DeveloperAddFeature/DeveloperAddFeature.xml]: Update filenames to match style guide and coding standards:

yournewapp.cpp
SupportClass.cpp
...

**Review by jessemapel — APPROVED** [review]

**Pull request merged**